sdk/storage/Azure.Storage.Blobs/src/PageBlobWriteStream.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Storage.Blobs.Models;
+using Azure.Storage.Blobs.Specialized;
+using Azure.Storage.Shared;
+
+namespace Azure.Storage.Blobs
+{
+    internal class PageBlobWriteStream : StorageWriteStream
+    {
+        private readonly PageBlobClient _pageBlobClient;
+        private readonly PageBlobRequestConditions _conditions;
+        private long _writeIndex;
+
+        public PageBlobWriteStream(
+            PageBlobClient pageBlobClient,
+            long bufferSize,
+            long position,
+            PageBlobRequestConditions conditions,
+            IProgress<long> progressHandler) : base(
+                position,
+                bufferSize,
+                progressHandler)
+        {
+            ValidateBufferSize(bufferSize);
+            ValidatePosition(position);
+            _pageBlobClient = pageBlobClient;
+            _conditions = conditions ?? new PageBlobRequestConditions();
+            _writeIndex = position;
+        }
+
+        protected override async Task WriteInternal(
+            byte[] buffer,
+            int offset,
+            int count,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            ValidateWriteParameters(buffer, offset, count);
+            int remaining = count;
+
+            // New bytes will fit in the buffer.
+            if (count <= _bufferSize - _buffer.Position)
+            {
+                await WriteToBufferInternal(buffer, offset, count, async, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                // We need a multiple of 512 to flush.
+                if (_buffer.Length % Constants.Blob.Page.PageSize != 0)
+                {
+                    int bytesToWrite = (int)(Constants.Blob.Page.PageSize - _buffer.Length % Constants.Blob.Page.PageSize);
+                    await WriteToBufferInternal(buffer, offset, bytesToWrite, async, cancellationToken).ConfigureAwait(false);
+                    remaining -= bytesToWrite;
+                    offset += bytesToWrite;
+                }
```

Comment:
This is the only part that's different than the base class?  Could we split out another virtual method that we can stick this into so we don't have to duplicate the base class method here?

Reply:
How would this work?  Pass in the method as a lambda from `PageBlobClient.OpenWrite()`?